ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -332,6 +334,10 @@ run ChainDB{..} internal registry varCurSlot = \case
           intScheduledChainSelection internal slot
       addBlock blk
 
+    giveWithEq :: a -> m (WithEq a)
+    giveWithEq a =
+      fmap (`WithEq` a) $ atomically $ updateTVar varNextId $ \i -> (succ i, i)
```

Comment:
Move this to `WithEq` and use a global counter (`StrictTVar`).

Reply:
Unfortunately, we don't have an `unsafePerformIO` (needed to created the global counter) that we can use in the simulator and the real IO monad.

So we're sticking with the current approach.